### PR TITLE
[wip] Fix invalid Kubernetes pod names

### PR DESCRIPTION
**Pull request by mayitbeegh**
# TODO
Update go.mod

# TL;DR
When running task execution, the propeller attempt to use the unvetted generated name for the execution as the pod name which creates errors on pod creation (e.g. `uw51wxgz3g-orgflyteexamplesHelloWorldTask-0`). The capitalized letters violates the Kubernetes' pod name restriction which is based on [the DNS label standard as defined in RFC-1123](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/). This PR attempts to address this problem (and future problems) by adding validation and conversion after the metadata (including the name) get added to the pod request object.
https://github.com/flyteorg/flytepropeller/blob/a1d959bbe20026145d27597f34501e59c9d6dbfd/pkg/controller/nodes/task/k8s/plugin_manager.go#L108 this overrides any changes made on the plugin side.

# Before:
`[Invalid] failed to create resource, caused by: Pod \"uw51wxgz3g-orgflyteexamplesHelloWorldTask-0\" is invalid: metadata.name: Invalid value: \"uw51wxgz3g-orgflyteexamplesHelloWorldTask-0\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character`

# After:
`{"json":{"exec_id":"yftcgyyovn","node":"orgflyteexamplesGreetTask","ns":"flytetester-development","res_ver":"271340","routine":"worker-1","src":"plugin_manager.go:211","tasktype":"java-task","wf":"flytetester:development:.flytegen.org.flyte.examples.GreetTask"},"level":"info","msg":"Creating Object: Type:[/v1, Kind=pod], Object:[flytetester-development/yftcgyyovn-orgflyteexamples-greet-task-0]","ts":"2021-09-17T22:29:26Z"}
{"json":{"exec_id":"yftcgyyovn","node":"orgflyteexamplesGreetTask","ns":"flytetester-development","res_ver":"271340","routine":"worker-1","src":"plugin_manager.go:185","tasktype":"java-task","wf":"flytetester:development:.flytegen.org.flyte.examples.GreetTask"},"level":"info","msg":"The resource requirement for creating Pod [flytetester-development/yftcgyyovn-orgflyteexamples-greet-task-0] is [[{[cpu]: [100m]} {[memory]: [200Mi]}]]\n","ts":"2021-09-17T22:29:26Z"}`

Also see https://github.com/flyteorg/flyteplugins/pull/211

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1172

### CODE REVIEW

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 1]
```diff
@@ -115,6 +117,13 @@ func (e *PluginManager) AddObjectMetadata(taskCtx pluginsCore.TaskExecutionMetad
 		f := append(o.GetFinalizers(), finalizer)
 		o.SetFinalizers(f)
 	}
+
+	pod, casted := o.(*v1.Pod)
+	if casted {
+		if errs := validation.IsDNS1123Label(pod.Name); len(errs) > 0 {
+			pod.Name = utils.ConvertToDNS1123CompatibleString(pod.Name)
+		}
+	}
```
Comment (pmahindrakar-oss) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:126]: What if casted is false. Should this be raising an error or is it expected that its always a v1.Pod castable object

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 2]
```diff
@@ -115,6 +117,13 @@ func (e *PluginManager) AddObjectMetadata(taskCtx pluginsCore.TaskExecutionMetad
 		f := append(o.GetFinalizers(), finalizer)
 		o.SetFinalizers(f)
 	}
+
+	pod, casted := o.(*v1.Pod)
+	if casted {
+		if errs := validation.IsDNS1123Label(pod.Name); len(errs) > 0 {
+			pod.Name = utils.ConvertToDNS1123CompatibleString(pod.Name)
```
Comment (pmahindrakar-oss) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:124]: Instead of two different function one for validation and one for conversion which then have to maintained for compatability if anything changes in one , is it ok if we have one function which validates and if its not compliant converts to the compatible format

Comment (mayitbeegh) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:124]: This is intended for fixing the pod name. I don't know if other request objects want the same treatment e.g. https://github.com/flyteorg/flyteplugins/blob/d0a6ee24404c73b2db6fe35dd59fd6a9e285f755/go/tasks/plugins/k8s/sagemaker/builtin_training.go#L37 TrainingJob.

File: pkg/controller/nodes/task/k8s/plugin_manager.go (hunk 1 above)
Comment (mayitbeegh) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:126]: This is intended for fixing the pod name. I don't know if other request objects want the same treatment e.g. https://github.com/flyteorg/flyteplugins/blob/d0a6ee24404c73b2db6fe35dd59fd6a9e285f755/go/tasks/plugins/k8s/sagemaker/builtin_training.go#L37 TrainingJob.

**Review by pmahindrakar-oss — APPROVED** [review]

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 3]
```diff
@@ -115,6 +117,13 @@ func (e *PluginManager) AddObjectMetadata(taskCtx pluginsCore.TaskExecutionMetad
 		f := append(o.GetFinalizers(), finalizer)
 		o.SetFinalizers(f)
 	}
+
+	pod, casted := o.(*v1.Pod)
+	if casted {
+		if errs := validation.IsDNS1123Label(pod.Name); len(errs) > 0 {
```
[the remark this answers is not in the document]
Comment (mayitbeegh) [reply pkg/controller/nodes/task/k8s/plugin_manager.go:123]: How would you do that? https://github.com/flyteorg/flytepropeller/blob/cdc1aff0a06ec266541978d9e9030a23fbe0b1d1/pkg/controller/nodes/task/k8s/plugin_manager.go#L263 
The client object's name `o.SetName(taskCtx.GetTaskExecutionID().GetGeneratedName())` gets invoked in `Abort`, `CheckResourcePhase`, `Finalize`, `LaunchResource`, and their names have to match. `TaskExecutionMetadata` and `TaskExecutionID` are both interfaces. I guess it's possible to implement `GetTaskExecutionID()` and `GetGeneratedName()` with the name conversion.

File: pkg/controller/nodes/task/k8s/utils.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+var dns1123InvalidRegex = regexp.MustCompile("[^-a-z0-9]")
+var camelCaseRegex = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+const maxUniqueIDLength = 20
+
+func ConvertToDNS1123CompatibleString(name string) string {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/utils.go:17]: Mind adding docs to the public APIs... ?

File: pkg/controller/nodes/task/k8s/utils.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		if err == nil {
+			name = name[:validation.DNS1123LabelMaxLength-maxUniqueIDLength-1] + "-" + fixedLengthID
+		} else {
+			name = name[:validation.DNS1123LabelMaxLength]
+		}
+	}
+	return name
+}
+
+func ConvertCamelCaseToKebabCase(name string) string {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/utils.go:33]: The other reason to keep these APIs in pluginsmachinary as you initially did is to be able to use them here:
https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/array/k8s/task.go#L115

Which does not use PluginManager to kick off pods... and will need the same handling... I believe

**Review by EngHabu — COMMENTED** [review]
The other place to conside

File: pkg/controller/nodes/task/k8s/utils.go (hunk 5 above)
Comment (mayitbeegh) [reply pkg/controller/nodes/task/k8s/utils.go:33]: I moved them here because of `encoder.go`. Do you want `encoder.go` moved to flyteplugins as well?

Comment (EngHabu) [reply pkg/controller/nodes/task/k8s/utils.go:33]: ic... yeah sure... maybe actually make a better package name in pluginmachinary for this? `encoding` or something?
